File: core/src/main/java/org/glowroot/jvm/HeapHistogram.java
/*
 * Copyright 2013 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.glowroot.jvm;

import java.io.IOException;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonGenerator;
import com.google.common.collect.Maps;
import com.google.common.collect.Ordering;
import com.google.common.io.CharStreams;
import org.objectweb.asm.Type;

/**
 * @author Trask Stalnaker
 * @since 0.5
 */
class HeapHistogram {

    private static final JsonFactory jsonFactory = new JsonFactory();

    private final Map<String, ClassInfo> classInfos = Maps.newHashMap();
    private long totalBytes;
    private long totalCount;

    void addItem(String className, long bytes, long count) {
        if (className.charAt(0) == '[') {
            className = Type.getType(className).getClassName();
        }
        ClassInfo classInfo = classInfos.get(className);
        if (classInfo == null) {
            classInfos.put(className, new ClassInfo(className, bytes, count));
        } else {
            classInfo.update(bytes, count);
        }
    }

    void setTotalBytes(long totalBytes) {
        this.totalBytes = totalBytes;
    }

    void setTotalCount(long totalCount) {
        this.totalCount = totalCount;
    }

    String toJson() throws IOException {
        StringBuilder sb = new StringBuilder();
        JsonGenerator jg = jsonFactory.createGenerator(CharStreams.asWriter(sb));
        jg.writeStartObject();
        jg.writeArrayFieldStart("items");
        List<ClassInfo> sortedClassInfos = ClassInfo.byBytesDesc.sortedCopy(classInfos.values());
        for (ClassInfo classInfo : sortedClassInfos) {
            jg.writeStartObject();
            jg.writeStringField("className", classInfo.getClassName());
            jg.writeNumberField("bytes", classInfo.getBytes());
            jg.writeNumberField("count", classInfo.getCount());
            jg.writeEndObject();
        }
        jg.writeEndArray();
        jg.writeNumberField("totalBytes", totalBytes);
        jg.writeNumberField("totalCount", totalCount);
        jg.writeEndObject();
        jg.close();
        return sb.toString();
    }

    private static class ClassInfo {

        private static final Ordering<ClassInfo> byBytesDesc = new Ordering<ClassInfo>() {
            @Override
            public int compare(ClassInfo left, ClassInfo right) {
                return Long.signum(right.bytes - left.bytes);
            }
        };

        private final String className;
        private long bytes;
        private long count;

        private ClassInfo(String className, long bytes, long count) {
            this.className = className;
            this.bytes = bytes;
            this.count = count;
        }

        private void update(long bytes, long count) {
            this.bytes += bytes;
            this.count += count;
        }

        private String getClassName() {
            return className;
        }

        private long getBytes() {
            return bytes;
        }

        private long getCount() {
            return count;
        }
    }
}

File: core/app/scripts/controllers/jvm/heap-histogram.js
/*
 * Copyright 2013-2014 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* global glowroot, angular, $ */

glowroot.controller('JvmHeapHistogramCtrl', [
  '$scope',
  '$http',
  '$q',
  '$timeout',
  'httpErrors',
  function ($scope, $http, $q, $timeout, httpErrors) {
    var sortAttribute = 'bytes';
    var sortDesc = true;

    // this is used to calculate bar width under class name representing the proportion of total bytes
    var maxBytes;

    $scope.classNameBarWidth = function (bytes) {
      return (bytes / maxBytes) * 100 + '%';
    };

    function sort() {
      if (sortAttribute === 'className') {
        // text sort
        $scope.histogram.items.sort(function (item1, item2) {
          var s1 = item1.className.toLowerCase();
          var s2 = item2.className.toLowerCase();
          var compare = s1 < s2 ? -1 : s1 > s2 ? 1 : 0;
          return sortDesc ? -compare : compare;
        });
      } else {
        // numeric sort
        $scope.histogram.items.sort(function (item1, item2) {
          var compare = item1[sortAttribute] - item2[sortAttribute];
          return sortDesc ? -compare : compare;
        });
      }
    }

    $scope.sortOn = function (attributeName) {
      if (sortAttribute === attributeName) {
        sortDesc = !sortDesc;
      } else {
        sortAttribute = attributeName;
        sortDesc = true;
      }
      sort();
      applyFilter();
    };

    $scope.refresh = function (deferred) {
      $http.get('backend/jvm/heap-histogram')
          .success(function (data) {
            $scope.loaded = true;
            $scope.histogram = data;
            if (sortAttribute !== 'bytes' || !sortDesc) {
              // data needs to be sorted
              sort();
            }
            applyFilter();
            if (deferred) {
              deferred.resolve('Complete');
            }
          })
          .error(httpErrors.handler($scope, deferred));
    };

    $scope.exportAsCsv = function (deferred) {
      var csv = 'Class name,Bytes,Count<br>';
      angular.forEach($scope.histogram.items, function (item) {
        csv += item.className + ',' + item.bytes + ',' + item.count + '<br>';
      });
      var csvWindow = window.open();
      $(csvWindow.document.body).html(csv);
      deferred.resolve();
    };

    $scope.filterComparatorOptions = [
      {
        display: 'Contains',
        value: 'contains'
      },
      {
        display: 'Begins with',
        value: 'begins'
      },
      {
        display: 'Ends with',
        value: 'ends'
      }
    ];
    $scope.filterComparator = 'contains';
    $scope.filterValue = '';
    $scope.filterLimit = 200;

    function applyFilter() {
      if ($scope.displayedItems === undefined) {
        $scope.displayedItems = [];
      }
      $scope.displayedItems.length = 0;
      $scope.limitApplied = false;
      $scope.filteredTotalBytes = 0;
      $scope.filteredTotalCount = 0;
      if ($scope.filterValue === '') {
        // optimization
        $scope.displayedItems = $scope.histogram.items.slice(0, $scope.filterLimit);
        $scope.limitApplied = ($scope.histogram.items.length > $scope.filterLimit);
        $scope.filteredTotalBytes = $scope.histogram.totalBytes;
        $scope.filteredTotalCount = $scope.histogram.totalCount;
        maxBytes = 0;
        angular.forEach($scope.displayedItems, function (item) {
          maxBytes = Math.max(maxBytes, item.bytes);
        });
        return;
      }
      var items = $scope.histogram.items;
      maxBytes = 0;
      for (var i = 0; i < items.length; i++) {
        var item = items[i];
        if (matchesFilter(item.className)) {
          if ($scope.displayedItems.length < $scope.filterLimit) {
            $scope.displayedItems.push(item);
            maxBytes = Math.max(maxBytes, item.bytes);
          } else {
            $scope.limitApplied = true;
          }
          $scope.filteredTotalBytes += item.bytes;
          $scope.filteredTotalCount += item.count;
        }
      }
    }

    function matchesFilter(className) {
      if ($scope.filterComparator === 'begins') {
        return className.indexOf($scope.filterValue) === 0;
      }
      if ($scope.filterComparator === 'ends') {
        return className.indexOf($scope.filterValue, className.length - $scope.filterValue.length) !== -1;
      }
      // contains
      return className.indexOf($scope.filterValue) !== -1;
    }

    $scope.$watch('filterComparator', function (newValue) {
      if ($scope.displayedItems === undefined) {
        // histogram hasn't loaded yet
        return;
      }
      applyFilter();
    });

    $scope.$watch('filterValue', function (newValue) {
      if ($scope.displayedItems === undefined) {
        // histogram hasn't loaded yet
        return;
      }
      applyFilter();
    });

    $scope.$watch('filterLimit', function (newValue) {
      if ($scope.displayedItems === undefined) {
        // histogram hasn't loaded yet
        return;
      }
      applyFilter();
    });

    $scope.refresh();
  }
]);
